README.rst
=========
wrapalloc
=========

.. contents::
.. sectnum::

Overview
--------

wrapalloc is a tool that intercepts calls to the libc memory
management routines. This allows it:

- perform checks on:

  - *overruns*
    (where the calling program writes beyond the end of the memory it
    has allocated).

  - *underruns*
    (where the calling program writes before the start of
    the memory it has allocated).

- display memory management statistics.
- arrange for memory to be initialised with a particular bit pattern.
- arrange for memory to be returned to the system after first
  overwriting it with a particular bit pattern.

wrapalloc is similar to ``dmalloc(1)`` and ``electric-fence(efence(3))``,
but it's lighter-weight (and less clever).

Design
------

When the user allocates a block of say 16 bytes of memory like this::

    char *p = malloc (16);

... what actually happens is::

     +---------+------------+--------+-------------+
     | MCB     | pre-buffer | memory | post-buffer |
     +---------+------------+--------+-------------+
      .memory               0 ..... 15            
         |     ^            ^                      ^
         |     |            |                      |
         +-----|------------^                      |
               |            |                      |
      .begin --+            p                      |
                                                   |
      .end ----------------------------------------+

      .request_size ------->{________}

      .total_size
          |
     +----+
     V
     {_____________________________________________}


The user is returned the address of a block of the size they specified.
However, we also allocate a pre- and post- buffer or "margin" either
side and immediately adjacent to the chunk of space requested.

Immediately before the pre-buffer is the MCB or "``MemoryCtlBlock``", a
structure used for recording information about the allocation and
buffers/margins. The MCB has a number of elements useful for tracking
the original allocation including:

- ``.memory`` which points to the memory block returned to the user.
- ``.begin`` which points to the start of the pre-buffer.
- ``.end`` which points to the end of the post-buffer.
- ``.request_size`` which records the size of the block of memory the
  user requested.
- ``.total_size`` which records the actual amount of memory allocated
  (requested size + pre-buffer size + post-buffer size + MCB).

If the program either overruns or underruns such that either the pre- or
post-buffers get modified, this is detected both when ``realloc(3)`` is
called, and when ``free(3)`` is called.

The size of the pre- and post-buffers are configurable using environment
variables, either independently::

  export WRAP_ALLOC_PRE_BORDER=<size>
  export WRAP_ALLOC_POST_BORDER=<size>

... or together::

  export WRAP_ALLOC_BORDER=<size>

Where "``<size>``" is a value in bytes. Note that the smaller the
border sizes, the lower the likelihood of detecting buffer over- and
under-runs.

Notes: 

The library is built using a linker trick [#linker-trick]_ to redirect
(interpose) all calls to the standard memory allocation routines with
the internal library versions.

Environment Variables
---------------------

Variables are used to modify the behaviour of this library. Note that
internally, those variables are queried *as late as possible* since
loading all recognised environment variables at library load time is
*NOT* useful since it precludes the caller from being able to modify the
behaviour (since the caller will run after library load time).

See the man page for the complete list of available environment
variables to control the behaviour of this library.

Limitations
-----------

- Won't work correctly for threaded apps (due to simplistic setting of 'caller').

- Although the code handles it, the library cannot be built to use the
  dynamic linker rather than the linker trick since the dynamic linker
  routines allocate memory, resulting in recursive execution followed
  by a crash.

Running
-------

Pre-requisites
~~~~~~~~~~~~~~

Make sure you disable all glibc/glib memory checkers by setting the
following prior to running the application you want `wrapalloc` to
check::

    # disable (e)glibc checking
    export MALLOC_CHECK_=0

    # disable glib checking
    export G_DEBUG=gc-friendly,resident-modules
    export G_SLICE=always-malloc

Playing with the tests
~~~~~~~~~~~~~~~~~~~~~~

Forcing an underrun
...................

Run the test program with ``LD_PRELOAD`` to force a write before the
buffer the test program allocates::

  $ (LD_PRELOAD=$PWD/src/.libs/libwrapalloc.so WRAP_ALLOC_LOGFILE=wrap-alloc.log ./tests/test_wrapalloc underrun)
  $ cat wrap-alloc.log 
  ERROR: underrun - expected fill byte 0x0 got 0x55 (1 byte before
  beginning of user memory 0x7fbc1145f058 of size 6)
  ERROR: damaged pre-border:
  INFO: caught SIGSEGV - aborting

Forcing an overrun
...................

Run the test program with ``LD_PRELOAD`` to force a write beyond the
buffer the test program allocates::

  $ (LD_PRELOAD=$PWD/src/.libs/libwrapalloc.so WRAP_ALLOC_LOGFILE=wrap-alloc.log ./tests/test_wrapalloc overrun)
  $ cat wrap-alloc.log 
  ERROR: overrun - expected fill byte 0x0 got 0x4f (1 byte beyond end of
  user memory 0x7ff19373c058 of size 6)
  ERROR: damaged post-border:
  INFO: caught SIGSEGV - aborting

Show the damaged buffer
~~~~~~~~~~~~~~~~~~~~~~~

By increasing the debug level, you can get a dump of the post buffer to
see exactly how the program has overwritten the memory::

  $ (LD_PRELOAD=$PWD/src/.libs/libwrapalloc.so WRAP_ALLOC_DEBUG=3 WRAP_ALLOC_LOGFILE=wrap-alloc.log WRAP_ALLOC_BORDER=8 ./tests/test_wrapalloc overrun)
  $ cat wrap-alloc.log 
  wrap-alloc:pid=5960:ppid=2928:file=wrap_alloc.c:line=1036:func=wa_init:DEBUG: --------------------
  wrap-alloc:pid=5960:ppid=2928:file=wrap_alloc.c:line=1038:func=wa_init:DEBUG: wrap-alloc version 0.1
  wrap-alloc:pid=5960:ppid=2928:file=wrap_alloc.c:line=1039:func=wa_init:DEBUG: 
  wrap-alloc:pid=5960:ppid=2928:file=wrap_alloc.c:line=1041:func=wa_init:DEBUG: build date: Jun  1 2015 at 20:17:22
  wrap-alloc:pid=5960:ppid=2928:file=wrap_alloc.c:line=1043:func=wa_init:DEBUG: build type:
  wrap-alloc:pid=5960:ppid=2928:file=wrap_alloc.c:line=1050:func=wa_init:DEBUG:   - mmap
  wrap-alloc:pid=5960:ppid=2928:file=wrap_alloc.c:line=1061:func=wa_init:DEBUG: functions:
  wrap-alloc:pid=5960:ppid=2928:file=wrap_alloc.c:line=1063:func=wa_init:DEBUG:   __wa_wrap_malloc=0x7fedd9787cbf
  wrap-alloc:pid=5960:ppid=2928:file=wrap_alloc.c:line=1064:func=wa_init:DEBUG:   __real_malloc=0x7fedd9786310
  wrap-alloc:pid=5960:ppid=2928:file=wrap_alloc.c:line=1066:func=wa_init:DEBUG:   __wa_wrap_calloc=0x7fedd9787d55
  wrap-alloc:pid=5960:ppid=2928:file=wrap_alloc.c:line=1067:func=wa_init:DEBUG:   __real_calloc=0x7fedd978632a
  wrap-alloc:pid=5960:ppid=2928:file=wrap_alloc.c:line=1069:func=wa_init:DEBUG:   __wa_wrap_realloc=0x7fedd9787ded
  wrap-alloc:pid=5960:ppid=2928:file=wrap_alloc.c:line=1070:func=wa_init:DEBUG:   __real_realloc=0x7fedd978634f
  wrap-alloc:pid=5960:ppid=2928:file=wrap_alloc.c:line=1079:func=wa_init:DEBUG:   __wa_wrap_free=0x7fedd9787ff5
  wrap-alloc:pid=5960:ppid=2928:file=wrap_alloc.c:line=1080:func=wa_init:DEBUG:   __real_free=0x7fedd9786374
  wrap-alloc:pid=5960:ppid=2928:file=wrap_alloc.c:line=1082:func=wa_init:DEBUG:   __cyg_profile_func_enter=0x7fedd94cfd40
  wrap-alloc:pid=5960:ppid=2928:file=wrap_alloc.c:line=1083:func=wa_init:DEBUG:   __cyg_profile_func_exit=0x7fedd94cfd40
  wrap-alloc:pid=5960:ppid=2928:file=wrap_alloc.c:line=1085:func=wa_init:DEBUG:   malloc=0x7fedd9786310
  wrap-alloc:pid=5960:ppid=2928:file=wrap_alloc.c:line=1086:func=wa_init:DEBUG:   calloc=0x7fedd978632a
  wrap-alloc:pid=5960:ppid=2928:file=wrap_alloc.c:line=1087:func=wa_init:DEBUG:   realloc=0x7fedd978634f
  wrap-alloc:pid=5960:ppid=2928:file=wrap_alloc.c:line=1088:func=wa_init:DEBUG:   free=0x7fedd9786374
  wrap-alloc:pid=5960:ppid=2928:file=wrap_alloc.c:line=1091:func=wa_init:DEBUG:    dlopen=0x7fedd91b7030
  wrap-alloc:pid=5960:ppid=2928:file=wrap_alloc.c:line=1092:func=wa_init:DEBUG:    dlsym=0x7fedd91b7100
  wrap-alloc:pid=5960:ppid=2928:file=wrap_alloc.c:line=1093:func=wa_init:DEBUG:    dlerror=0x7fedd91b7370
  wrap-alloc:pid=5960:ppid=2928:file=wrap_alloc.c:line=1094:func=wa_init:DEBUG:    dlclose=0x7fedd91b70a0
  wrap-alloc:pid=5960:ppid=2928:file=wrap_alloc.c:line=1098:func=wa_init:DEBUG:    mmap=0x7fedd94bb720
  wrap-alloc:pid=5960:ppid=2928:file=wrap_alloc.c:line=1099:func=wa_init:DEBUG:    munmap=0x7fedd94bb750
  wrap-alloc:pid=5960:ppid=2928:file=wrap_alloc.c:line=1102:func=wa_init:DEBUG: option settings:
  wrap-alloc:pid=5960:ppid=2928:file=wrap_alloc.c:line=1104:func=wa_init:DEBUG:   free: enabled
  wrap-alloc:pid=5960:ppid=2928:file=wrap_alloc.c:line=1107:func=wa_init:DEBUG:   pre buffer size: 8
  wrap-alloc:pid=5960:ppid=2928:file=wrap_alloc.c:line=1109:func=wa_init:DEBUG:   post buffer size: 8
  wrap-alloc:pid=5960:ppid=2928:file=wrap_alloc.c:line=1112:func=wa_init:DEBUG:   pre buffer fill byte: 0x0
  wrap-alloc:pid=5960:ppid=2928:file=wrap_alloc.c:line=1114:func=wa_init:DEBUG:   post buffer fill byte: 0x0
  wrap-alloc:pid=5960:ppid=2928:file=wrap_alloc.c:line=1115:func=wa_init:DEBUG:   alloc byte: 0x0
  wrap-alloc:pid=5960:ppid=2928:file=wrap_alloc.c:line=1116:func=wa_init:DEBUG:   free byte: 0x0
  wrap-alloc:pid=5960:ppid=2928:file=wrap_alloc.c:line=1117:func=wa_init:DEBUG:   debug: 0x3
  wrap-alloc:pid=5960:ppid=2928:file=wrap_alloc.c:line=1118:func=wa_init:DEBUG: --------------------
  wrap-alloc:pid=5960:ppid=2928:file=wrap_alloc.c:line=811:func=__wa_wrap_malloc:DEBUG: caller requested allocation of 6 bytes
  wrap-alloc:pid=5960:ppid=2928:file=wrap_alloc.c:line=638:func=__wa_new_mem_block:DEBUG: __wa_new_mem_block called with size=6
  wrap-alloc:pid=5960:ppid=2928:file=wrap_alloc.c:line=592:func=wa_show_ctl_block:DEBUG: MemoryCtlBlock=0x7fedd9bac000
  wrap-alloc:pid=5960:ppid=2928:file=wrap_alloc.c:line=593:func=wa_show_ctl_block:DEBUG:   pre_border_size=8
  wrap-alloc:pid=5960:ppid=2928:file=wrap_alloc.c:line=594:func=wa_show_ctl_block:DEBUG:   post_border_size=8
  wrap-alloc:pid=5960:ppid=2928:file=wrap_alloc.c:line=595:func=wa_show_ctl_block:DEBUG:   eye_catcher='WACTLBK'
  wrap-alloc:pid=5960:ppid=2928:file=wrap_alloc.c:line=596:func=wa_show_ctl_block:DEBUG:   memory=0x7fedd9bac060
  wrap-alloc:pid=5960:ppid=2928:file=wrap_alloc.c:line=597:func=wa_show_ctl_block:DEBUG:   begin=0x7fedd9bac058
  wrap-alloc:pid=5960:ppid=2928:file=wrap_alloc.c:line=598:func=wa_show_ctl_block:DEBUG:   end=0x7fedd9bac06e
  wrap-alloc:pid=5960:ppid=2928:file=wrap_alloc.c:line=599:func=wa_show_ctl_block:DEBUG:   request_size=6
  wrap-alloc:pid=5960:ppid=2928:file=wrap_alloc.c:line=600:func=wa_show_ctl_block:DEBUG:   total_size=110
  wrap-alloc:pid=5960:ppid=2928:file=wrap_alloc.c:line=608:func=wa_show_ctl_block:DEBUG:   call_time=1433187315.274992508
  wrap-alloc:pid=5960:ppid=2928:file=wrap_alloc.c:line=737:func=__wa_new_mem_block:DEBUG: filling user buffer with alloc byte value 0x0
  wrap-alloc:pid=5960:ppid=2928:file=wrap_alloc.c:line=741:func=__wa_new_mem_block:DEBUG: returning m->memory=0x7fedd9bac060
  wrap-alloc:pid=5960:ppid=2928:file=wrap_alloc.c:line=947:func=__wa_wrap_free:DEBUG: caller requested freeing of 6 bytes
  wrap-alloc:pid=5960:ppid=2928:file=wrap_alloc.c:line=568:func=wa_check_ctl_block:ERROR: overrun - expected fill byte 0x0 got 0x4f (1 byte beyond end of user memory 0x7fedd9bac060 of size 6)
  wrap-alloc:pid=5960:ppid=2928:file=wrap_alloc.c:line=570:func=wa_check_ctl_block:ERROR: damaged post-border:
  wrap-alloc:pid=5960:ppid=2928:file=wa_util.c:line=473:func=wa_tohex:DEBUG: 000000: 4f00 0000 0000 0000                      O.......
  wrap-alloc:pid=5960:ppid=2928:file=wrap_alloc.c:line=1334:func=wa_abort:INFO: caught SIGSEGV - aborting

Notes:

- The penultimate line above shows the bogus write - the ``test_wrapalloc`` program writes a "``O``" when requested to perform an over-run.
- The default buffer has been changed to only 8 bytes using the ``WRAP_ALLOC_BORDER`` variable to make the example above clearer.


Running a standard command using ``LD_PRELOAD``
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Here's an example of running "``sleep 1``" via ``LD_PRELOAD``::

  $ (LD_PRELOAD=$PWD/src/.libs/libwrapalloc.so WRAP_ALLOC_LOGFILE=wrap-alloc.log /bin/sleep 1)

Footnotes
---------

.. [#linker-trick]

   The linker trick is "``ld -Wl,--wrap=malloc,--wrap=calloc,--wrap=free ...``".



NEWS
0.1.0  2015-05-29
 
	* Initial public release.


tests/test_wrapalloc.c
/*--------------------------------------------------------------------
 * wrapalloc - basic test program.
 *
 * Copyright (C) 2012-2015 James Hunt <james.hunt@ubuntu.com>.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *--------------------------------------------------------------------
 */

/*--------------------------------------------------------------------
 * Description: Test program for wrapalloc that can cause an underrun or
 * overrun, but which is NOT linked with -lwrapalloc. This allows us to
 * test the LD_PRELOAD handling.
 *--------------------------------------------------------------------
 */
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <assert.h>
#include <errno.h>

#define WORD "hello"

char *program_name;

void
test_wrapalloc (const char *action)
{
    char   *string = WORD;
    size_t  len;

    assert (action);

    len = strlen (string);

    string = (char *)malloc (len+1);

    if (! string) {
        fprintf (stderr,
                "%s:%d: malloc failed - errno=%d\n",
                __func__, __LINE__, errno);
        exit (EXIT_FAILURE);
    }

    strncpy (string, WORD, len+1);

    if (! strcmp (action, "underrun")) {
        fprintf (stderr, "forcing an underrun\n");
        *(string-1) = 'U';
    } else if (! strcmp (action, "overrun")) { 
        fprintf (stderr, "forcing an overrun\n");
        *(string + len+1) = 'O';
    } else {
        fprintf (stderr, "No under/over-run requested\n");
    }

    free (string);
}

int main (int argc, char *argv[])
{
    const char *action;

    program_name = argv[0];

    if (argc != 2) {
        fprintf (stderr, "ERROR: usage: %s <underrun|overrun|normal>\n",
                program_name);
        exit (EXIT_FAILURE);
    }

    action = argv[1];
    test_wrapalloc (action);

    exit (EXIT_SUCCESS);
}


src/wa_list.c
/*--------------------------------------------------------------------
 * wrapalloc - generic list handling routines.
 *
 * Copyright (C) 2012-2015 James Hunt <james.hunt@ubuntu.com>.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *--------------------------------------------------------------------
 */

#include <wa_list.h>

static inline WMList *wa_list_cut (WMList *entry);

WMList *
wa_list_new (void)
{
    WMList *list;

    list = (WMList *)wa_get_memory (sizeof (WMList));
    if (! list)
        return NULL;

    wa_list_init (list);

    return list;
}

void
wa_list_init (WMList *entry)
{
    wa_assert (entry);

    entry->prev = entry->next = entry;
}

WMList *
wa_list_add (WMList *list,
        WMList *entry)
{
    wa_assert (list);
    wa_assert (entry);

    wa_list_cut (entry);

    entry->prev = list->prev;
    list->prev->next = entry;
    list->prev = entry;
    entry->next = list;

    return entry;
}

WMList *
wa_list_add_after (WMList *list,
        WMList *entry)
{
    wa_assert (list);
    wa_assert (entry);

    wa_list_cut (entry);

    entry->next = list->next;
    list->next->prev = entry;
    list->next = entry;
    entry->prev = list;

    return entry;
}

static inline WMList *
wa_list_cut (WMList *entry)
{
    wa_assert (entry);

    entry->prev->next = entry->next;
    entry->next->prev = entry->prev;

    return entry;
}

WMList *
wa_list_remove (WMList *entry)
{
    wa_assert (entry);

    wa_list_cut (entry);
    wa_list_init (entry);

    return entry;
}

int
wa_list_destroy (WMList *entry)
{
    wa_assert (entry);

    wa_list_cut (entry);

    return 0;
}


/**
 * wa_list_foreach:
 *
 * @list: list,
 * @len: pointer to int that will contain length of list,
 * @handler: optional function called for each list entry,
 * @data: optional data to pass to handler along with list entry.
 *
 * If @handler is NULL, list length will be returned in @len.
 * If @handler returns 1, @len will be set to the number of list entries
 * processed successfully up to that point.
 *
 * Returns 0 on success, or -1 if handler returns an error.
 **/
int
wa_list_foreach (const WMList *list, size_t *len,
        WMListHandler handler, void *data)
{
    int ret;

    wa_assert (list);

    *len = 0;

    WA_LIST_FOREACH (list, iter) {
        if (handler) {
            ret = handler (iter, data);
            if (!ret) return -1;
        }
        ++*len;
    }

    return 0;
}


src/wa_list.h
/*--------------------------------------------------------------------
 * wrapalloc - generic list handling header.
 *
 * Copyright (C) 2012-2015 James Hunt <james.hunt@ubuntu.com>.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *--------------------------------------------------------------------
 */

#ifndef _WM_LIST
#define _WM_LIST

#include <stdio.h>

#include <wa_util.h>

typedef struct wa_list {
    struct wa_list *prev, *next;
} WMList;

#define WA_LIST_FOREACH(list, iter)                                    \
    for (WMList *iter = (list)->next; iter != (list); iter = iter->next)

#define WA_LIST_EMPTY(list) (((list)->prev == (list))          \
        && ((list)->next) == (list))


typedef int (*WMListHandler) (WMList *entry, void *data);

WMList * wa_list_new (void);
void     wa_list_init (WMList *entry);
WMList * wa_list_add (WMList *list, WMList *entry);
WMList * wa_list_add_after (WMList *list, WMList *entry);
WMList * wa_list_remove (WMList *entry);
int      wa_list_destroy (WMList *entry);
int      wa_list_foreach(const WMList *list,
         size_t *len, WMListHandler handler, void *data);

#endif /* _WM_LIST */


src/wa_util.h
/*--------------------------------------------------------------------
 * wrapalloc - utility routines header.
 *
 * Copyright (C) 2011-2015 James Hunt <james.hunt@ubuntu.com>.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *--------------------------------------------------------------------
 */

#ifndef _WA_UTIL
#define _WA_UTIL

#ifndef TRUE
#define TRUE 1
#endif

#ifndef FALSE
#define FALSE (!TRUE)
#endif

#if (defined WA_MAIN) || defined (WA_TESTS)
#define WA_PRIVATE /*static*/
#else /* ! (WA_MAIN | WA_TESTS) */
#define WA_PRIVATE extern
#endif /* (WA_MAIN | WA_TESTS) */

/********************************************************************/

struct wa_signal_map {
    int    signum;
    char  *signame;
};

extern struct wa_signal_map wa_signal_map[];

/********************************************************************/
/* prototypes */

WA_PRIVATE void *
__attribute ((malloc, warn_unused_result, no_instrument_function, unused))
wa_get_memory (size_t size);

WA_PRIVATE void
wa_signal_handler (int signum);

WA_PRIVATE void
wa_setup_signals (void);

/* assert(3) can call malloc(3), so we have to do the job ourselves.
 *
 * XXX: Note the __no_return__ attribute to keep scan-build from
 * claiming code could deref a NULL Pointer after a call to wa_assert().
 */
WA_PRIVATE void
wa_assert_fail (const char *file,
        int line,
        const char *function,
        const char *expr)
    __attribute__((__noreturn__));

WA_PRIVATE void
_wa_log_msg (const char *file,
        int line,
        const char *func,
        const char *fmt, ...)
__attribute ((no_instrument_function));

WA_PRIVATE int
wa_address_valid (void *ptr);

int
wa_address_was_valid (void *ptr);

WA_PRIVATE void
wa_mcb_list_init (void);

WA_PRIVATE void
wa_address_list_init (void);

WA_PRIVATE void
wa_tohex (size_t bytes, const void *data);

WA_PRIVATE int
wa_show_printable (size_t bytes, const void *data, char *buffer);

WA_PRIVATE const char *
wa_signal_num_to_name (int signum);

WA_PRIVATE int
wa_signal_name_to_num (const char *signame);

/**************************************/

#ifdef USE_LD_PRELOAD

WA_PRIVATE void *(*__real_malloc)(size_t);
WA_PRIVATE void *(*__real_calloc)(size_t nmemb, size_t size);
WA_PRIVATE void *(*__real_realloc)(void *ptr, size_t size);
WA_PRIVATE void (*__real_free)(void *ptr);

#ifdef HAVE_ALLOCA
WA_PRIVATE void *(*__real_alloca)(size_t size);
#endif /* HAVE_ALLOCA */

#else /* ! USE_LD_PRELOAD */

extern void *__real_malloc (size_t size);
extern void *__real_calloc (size_t nmemb, size_t size);
extern void *__real_realloc (void *ptr, size_t size);
extern void __real_free (void *ptr);

#ifdef HAVE_ALLOCA
extern void *__real_alloca (size_t size);
#endif /* HAVE_ALLOCA */

#endif /* USE_LD_PRELOAD */

/********************************************************************/

#define APP_NAME    "wrap-alloc"
#define APP_VERSION "0.1"

#define WA_DELIMITER  "--------------------\n"

/* Value that marks an MemoryCtlBlock */
#define WA_EYE_CATCHER      "WACTLBK"

/* Length of EYE_CATCHER */
#define WA_EYE_CATCHER_LEN  (7+1)

/* size of buffer used by logger */
#define WA_BUFSIZE              1024
#define WA_LOG_BUFSIZE         WA_BUFSIZE

#define WA_DEFAULT_BUFFER_SIZE 4096

/* Behave like calloc(3) by default */
#ifndef DEFAULT_FILL_BYTE
  #define DEFAULT_FILL_BYTE  0x0
#endif

#define DEFAULT_ALLOC_FILL_BYTE DEFAULT_FILL_BYTE
#define DEFAULT_FREE_FILL_BYTE  DEFAULT_FILL_BYTE

/********************************************************************/
/* environment variables */

/* Enable debug output if set to any value */
#define WRAP_ALLOC_DEBUG_ENV        "WRAP_ALLOC_DEBUG"

/* Size of border buffer (in bytes) */
#define WRAP_ALLOC_BORDER_ENV   "WRAP_ALLOC_BORDER"

/* Size of pre buffer (in bytes).
 *
 * If not specified, uses WRAP_ALLOC_BORDER_ENV, or if that
 * isn't specified, uses the systems page size (atleast 4k).
 */
#define WRAP_ALLOC_PRE_BORDER_ENV   "WRAP_ALLOC_PRE_BORDER"

/* Size of post buffer (in bytes).
 *
 * If not specified, uses WRAP_ALLOC_BORDER_ENV, or if that
 * isn't specified, uses the systems page size (atleast 4k).
 */
#define WRAP_ALLOC_POST_BORDER_ENV  "WRAP_ALLOC_POST_BORDER"

/* Byte value used to fill pre and post buffers */
#define WRAP_ALLOC_FILL_ENV         "WRAP_ALLOC_FILL"

/* Byte value used to fill pre buffers.
 *
 * If not specified, uses WRAP_ALLOC_FILL_ENV, or if that isn't
 * specified, uses DEFAULT_FILL_BYTE.
 */
#define WRAP_ALLOC_PRE_FILL_ENV     "WRAP_ALLOC_PRE_FILL"

/* Byte value used to fill post buffers.
 *
 * If not specified, uses WRAP_ALLOC_FILL_ENV, or if that isn't
 * specified, uses DEFAULT_FILL_BYTE.
 */
#define WRAP_ALLOC_POST_FILL_ENV    "WRAP_ALLOC_POST_FILL"

/* Byte value use to fill memory area requested by user.
 * If not specified, value will be DEFAULT_ALLOC_FILL_BYTE.
 */
#define WRAP_ALLOC_ALLOC_BYTE_ENV   "WRAP_ALLOC_ALLOC_BYTE"

/* Byte value use to fill memory area requested by user just before
 * memory is passed to free(3).
 * If not specified, value will be DEFAULT_FREE_FILL_BYTE.
 */
#define WRAP_ALLOC_FREE_BYTE_ENV   "WRAP_ALLOC_FREE_BYTE"

/* If set, never free memory */
#define WRAP_ALLOC_DISABLE_FREE_ENV  "WRAP_ALLOC_DISABLE_FREE"

/* FIXME: we don't seem to display the addresses atm? */
/* If set, store all addresses in @address_list to aid in debugging */
#define WRAP_ALLOC_STORE_ALL_ADDRESSES_ENV "WRAP_ALLOC_STORE_ALL_ADDRESSES"

/* If set, install a SIGSEGV handler that will be agressively installed
 * and will pause the process.
 */
#define WRAP_ALLOC_SIGSEGV_HANDLER_ENV "WRAP_ALLOC_SIGSEGV_HANDLER"

#define WRAP_ALLOC_SEGV_ACTION_ENV "WRAP_ALLOC_SEGV_ACTION"

// FIXME: implement
#define WRAP_ALLOC_LOGFILE_ENV "WRAP_ALLOC_LOGFILE"

/********************************************************************/

/* Convert a "user pointer" as returned by (m|c)alloc to a
 * MemoryCtlBlock.
 */
#define wa_ptr_to_mcb(ptr) \
    (MemoryCtlBlock *) \
    ((byte *)ptr \
        - (sizeof (MemoryCtlBlock) \
        + wa_get_border_size (WA_BUFFER_TYPE_PRE)))

#define wa_mcb_to_pre_border(mcb) \
    ((byte *)(mcb)->begin)

#define wa_mcb_to_post_border(mcb) \
    (((byte *)(mcb)->end) - wa_get_border_size (WA_BUFFER_TYPE_POST))

#define wa_assert(expr) \
    ((expr)) \
        ? (void)0 \
        : wa_assert_fail (__FILE__, __LINE__, __func__, #expr)

/* Display an informational message */
#define wa_msg(...) \
    _wa_log_msg (__FILE__, __LINE__, __func__, "INFO: " __VA_ARGS__)

#define wa_warn(...) \
    _wa_log_msg (__FILE__, __LINE__, __func__, "WARNING: " __VA_ARGS__)

/* FIXME */
#if 0
#define wa_debug(...) \
    if (wa_debug_value) \
        _wa_log_msg (__FILE__, __LINE__, __func__, "DEBUG: " __VA_ARGS__)
#endif
#define wa_debug(...) \
    if (getenv (WRAP_ALLOC_DEBUG_ENV)) \
        _wa_log_msg (__FILE__, __LINE__, __func__, "DEBUG: " __VA_ARGS__)

#define wa_err(...) \
    _wa_log_msg (__FILE__, __LINE__, __func__, "ERROR: " __VA_ARGS__)

/* Return TRUE if sufficient data has been processed to display a line
 * of data.
 *
 * bytes: 0-based byte count.
 */
#define WA_HAVE_LINE_TO_DISPLAY(bytes) \
        (((bytes + 1) % WA_BYTES_PER_LINE) == 0)

#define WA_BYTES_PER_LINE 16

#define WA_TOHEX_OUTPUT_AND_CLEAR(buffer, p, ret, byte_count) \
    wa_debug ("%s\n", buffer); \
    memset (buffer, '\0', WA_BUFSIZE); \
    ret = byte_count = 0; \
    p = buffer

#define wa_signal_map_entry(s) \
    {s, #s}

/********************************************************************/

#ifdef USE_INSTRUMENT
/**
 * WA_IGNORE_WRAPPERS:
 *
 * Macro called when instrumenting to ensure we don't trace the wrapper
 * functions created by the linker.
 **/
#define WA_IGNORE_WRAPPERS() \
    if (func ==__wrap_malloc   || \
        func == __wrap_calloc  || \
        func == __wrap_realloc || \
        func == __wrap_alloca  || \
        func == __wrap_free)      \
            return
#endif /* USE_INSTRUMENT */

#endif /* _WA_UTIL */
